prometheusbp/promtest/spec.go

```diff
@@ -0,0 +1,80 @@
+package promtest
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+)
+
+var (
+	errPrefix         = errors.New("the prefix is not at the beginning of the metric name")
+	errLength         = errors.New("metric name should have a minimum of 3 parts, like prefix_name_suffix")
+	errCount          = errors.New("wrong metric count for prefix")
+	errPrometheusLint = errors.New("problem with Prometheus GatherAndLint")
+)
+
+// ValidateSpec validates that the Prometheus metrics being exposed conform to
+// the baseplate spec.
+func ValidateSpec(t *testing.T, metricPrefix string, wantMetricCount int) {
+	if err := validateSpec(metricPrefix, wantMetricCount); err != nil {
+		t.Error(err)
+	}
+}
```

Comment:
"wantMetricCount" is how many metrics there should be with the metricPrefix. I can add that to the doc string. 

I'm not sure what you mean "wantServiceCount" or even "wantServiceNames ...string".

Reply:
+1 on this sentiment, it becomes clear as soon as I scroll to the validation where it describes in the comment:
```
// validateName checks that:
// 1) the metric name has at least 3 parts separated by "_".
```
explaining why we pass 3 here
```
defer promtest.ValidateSpec(t, "thrift", 3)
```

Not sure how this can be written in a more expressive way. That said, I think those tests often serve as something that protects against any type of regressions so potentially not too important.

Reply:
ok, so if `wantMetricCount` is just "go to the spec and count how many metrics are expected", then isn't that already part of the spec? The caller doesn't need to provide that.  Any extra metrics that are not expected shuld result in an error, any missing metrics should result in an error.